casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -31,26 +31,13 @@ object ProtoUtil {
   /*
    * c is in the blockchain of b iff c == b or c is in the blockchain of the main parent of b
    */
-  def isInMainChain[F[_]: Monad](
+  def isInMainChain[F[_]: Sync](
       dag: BlockDagRepresentation[F],
       candidateMetadata: BlockMetadata,
       targetBlockHash: BlockHash
   ): F[Boolean] = {
-    import coop.rchain.catscontrib.Catscontrib.ToBooleanF
-    import cats.instances.option._
-
-    (candidateMetadata.blockHash == targetBlockHash).pure[F] ||^
-      dag.lookup(targetBlockHash).flatMap {
-        case Some(targetBlock) =>
-          if (targetBlock.blockNum <= candidateMetadata.blockNum) {
-            false.pure[F]
-          } else {
-            val mainParentOpt = targetBlock.parents.headOption
-            mainParentOpt.traverse(isInMainChain(dag, candidateMetadata, _)).map(_.getOrElse(false))
-          }
-        case None =>
-          false.pure[F]
-      }
+    val heightF: BlockHash => F[Long] = h => dag.lookupUnsafe(h).map(_.blockNum)
+    dag.isInMainChain(targetBlockHash, candidateMetadata.blockHash, heightF)
```

Comment:
isInMainChain is in DagReaderSyntax, but it requires heightF, so this function extracts height from BlockDagRepresentation.

Reply:
What makes this confusing is that DagReader is also a BlockDagRepresentation but it receives function to extract height which is available to this instance.

My point here is that it requires the whole BlockMetadata just to read the hash. But this can be cleanup later also.